### PR TITLE
Update channel topic when Test Day starts and ends

### PR DESCRIPTION
**Pull request by galgeek**
hi @whimboo !

with apologies for the untidy commit list, suspecting this requires re-forking the repo to fix, and not sure what that would do to existing pull requests

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -10,12 +10,15 @@ var ircServer = config.server,
       autoRejoin: config.autoRejoin,
     },
     client = new irc.Client(ircServer, nick, options),
+    botChannel = config.channels[0],
```
Comment (whimboo) [inline bot.js]: Just name it `channel`. It won't be anything else than the bot. :)

File: bot.js [hunk 2]
```diff
@@ -53,15 +56,23 @@ function checkTestDay() {
...
       testDay = true;
       resetData();
+      client.send('TOPIC', botChannel, topic);
     }
   }
 }
 
+client.addListener('topic', function (channel, channelTopic, nick) {
+  if (!testDay && (channel === botChannel)){
+    topic_backup = channelTopic; // save a non-Test Day topic to restore after Test Day
```
Comment (whimboo) [inline bot.js]: nit: Please move the comment to the line before.

**Comment by whimboo** [comment]
Also I think at the moment the bot waits for an activity in the channel before it updates the topic of the channel when the testday is about to start. Shouldn't this be handled with a timer event or such?

File: bot.js (hunk 1 above)
Comment (galgeek) [inline bot.js]: node irc uses `channel` throughout its documentation for function parameters, so it's a little confusing for anyone who's looking at that documentation to use `channel` as a separate variable name. I've made the change. I wonder what you'd prefer for the node irc function parameters, which I'm calling `channl` for now.

**Comment by galgeek** [comment]
hi @whimboo 

a biggish update, now using setTimeout to start and end test day, and other updates per your recent feedback

thank you!

File: bot.js (hunk 1 above)
Comment (whimboo) [inline bot.js]: By default we use an `a` prefix for function parameters across all of our repositories. It would be good to refactor the code appropriate, but lets move this out to another issue. This could even be a good first mentored bug for someone. Lets skip that for now.

File: bot.js [hunk 3]
```diff
@@ -51,19 +55,25 @@ function resetData() {
...
+    testDay = true;
+    resetData();
+    client.send('TOPIC', channel, topic);
+    timerID = setTimeout(checkTestDay, endTime - Date.now());
   }
 }
 
-client.addListener('join', function(channel, who){
-  checkTestDay();
+client.addListener('topic', function (channl, channlTopic, nick) {
```
Comment (whimboo) [inline bot.js]: Please revert this change to the parameter names. Stripping of chars should never be an option.

File: bot.js [hunk 4]
```diff
@@ -51,19 +55,25 @@ function resetData() {
...
+    timerID = 0;
   } else {
-    if ((Date.now() < endTime) && (Date.now() > startTime)){
-      testDay = true;
-      resetData();
-    }
+    testDay = true;
+    resetData();
+    client.send('TOPIC', channel, topic);
+    timerID = setTimeout(checkTestDay, endTime - Date.now());
```
Comment (whimboo) [inline bot.js]: I'm not really a friend of setting timers for a long timespan. It should be ok for now, but in the future we might want to have a period like timer, which checks for all different kinds of tasks each minute or so.

File: bot.js [hunk 5]
```diff
@@ -51,19 +55,25 @@ function resetData() {
 
 function checkTestDay() {
```
Comment (whimboo) [inline bot.js]: We should update the name of the method. It's not really a check anymore but more an `updateTestDayData`.

File: bot.js [hunk 6]
```diff
@@ -168,8 +176,8 @@ client.addListener('pm', function(from, message){ // private messages to bot
   if (testDay){
     if (message.search(':stop') === 0){
       if (testDayAdmins.indexOf(from) >= 0){
-        testDay = false;
-        endTime = Date.now();
+        clearTimeout(timerID);
```
Comment (whimboo) [inline bot.js]: Please do that in the checkTestDay method.

File: bot.js [hunk 7]
```diff
@@ -185,10 +193,19 @@ client.addListener('pm', function(from, message){ // private messages to bot
...
-        client.say(from, "Next test day's start is " + startTime);
-        client.say(from, "Next test day's end is " + endTime);
-        client.say(from, "Next test day's etherpad is " + etherpad);
-        client.say(from, "Next test day's topic is " + topic);
+        // if the start and end dates appear valid, set the test date
+        if ((endTime > startTime) && (startTime > Date.now())){
+          if (timerID !== 0){
+            clearTimeout(timerID);
+          }
+          timerID = setTimeout(checkTestDay, startTime - Date.now());
```
Comment (whimboo) [inline bot.js]: We cannot clear the timer here, especially if a testday is currently active.

File: bot.js (hunk 4 above)
Comment (whimboo) [inline bot.js]: Also keep in mind that this function is not called if the bot is joining IRC with an already set next testday. Currently we do not save the testday info, but that is something we have to do. Please file a new issue which can cover that.

File: bot.js (hunk 7 above)
Comment (galgeek) [inline bot.js]: The current code runs the :next command only when there's not a test day in progress.
Clearing the timer here allows us to reschedule a test day.

**Comment by galgeek** [comment]
Hi @whimboo 

I've updated per your comments, including adding issue #43 save and reload test day schedule across bot restarts.

Note that current code accepts the :next admin command only when a test day is not in progress, so I've made no changes to `clearTimeout` there.

File: bot.js (hunk 7 above)
Comment (whimboo) [inline bot.js]: Sorry, but I do not see any check for an active testday in the :next callback. You only check if the given times are valid and then reset the timer. This will kill the timer to reset the topic when the currently active testday is over. What you want instead is to check in updateTestdayData() if a next testday is scheduled, and set the timer appropriately.

**Comment by galgeek** [comment]
Hi @whimboo!

I've rebased and added test and qa config files.

I've left this version of the bot running using the test config, with a test Test Day scheduled for 6A-10A UTC. I'll check that it starts before I go to sleep, and maybe you'll have a chance to check that it's still running fine in the morning CET.

Thank you!

File: config.json.test [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+    "channels": ["#testdaybotTest"],
+    "server": "irc.mozilla.org",
+    "port": "6697",
+    "nick": "_TestDayBot",
+    "secure": "true",
+    "autoRejoin": "true",
+    "admins": ["ashughes","whimboo","galgeek","bara"],
+    "helpers": ["ashughes","whimboo","bara"]
+}
```
Comment (whimboo) [inline config.json.test]: Please remove that file from the patch. It's nice for testing but does not belong to the repository.

File: config.json.qa [hunk 9]
```diff
@@ -0,0 +1,10 @@
+{
+    "channels": ["#qa"],
+    "server": "irc.mozilla.org",
+    "port": "6697",
+    "nick": "_TestDayBot",
+    "secure": "true",
+    "autoRejoin": "true",
+    "admins": ["ashughes", "whimboo"],
+    "helpers": ["ashughes"]
+}
```
Comment (whimboo) [inline config.json.qa]: Same for this file.

File: bot.js [hunk 10]
```diff
@@ -195,8 +206,7 @@ client.addListener('pm', function(from, message) { // private messages to bot
         break;
       case ":stop":
         if (testDay) {
-          testDay = false;
```
Comment (whimboo) [inline bot.js]: I don't really like that this got moved into `updateTestDayData()` because it forces an early abort.

File: bot.js [hunk 11]
```diff
@@ -52,21 +56,30 @@ function resetData() {
...
+    if (timerID !== 0) {
+      clearTimeout(timerID);
+      timerID = 0;
     }
   } else {
-    if ((Date.now() < endTime) && (Date.now() > startTime)) {
-      testDay = true;
-      resetData();
-    }
+    testDay = true;
```
Comment (whimboo) [inline bot.js]: This changes the status of the testday now each time the method get called. :/ It's ok for now but in the future this method should really only check if any action needs to be done depending on the current state of the testday.

File: bot.js (hunk 10 above)
Comment (whimboo) [inline bot.js]: But that's necessary due to the current implementation of this method. See my above comment in what we should do later.

**Comment by galgeek** [comment]
Hi @whimboo!

I've removed the testing configs from the repo and updated .gitignore, and I've updated the PR.

I'm leaving any additional changes to updateTestDayData for later.

Thank you!

File: .gitignore [hunk 12]
```diff
@@ -1,3 +1,4 @@
 *.swp
 .DS_Store
-
+config.json.test
+config.json.qa
```
Comment (whimboo) [inline .gitignore]: Those are your own personal files on your machine. This is not for general use. So please remove.

**Comment by galgeek** [comment]
Hi, @whimboo!

As we noted in IRC, I've updated this PR, backing out the .gitignore edits.

Thank you!

Barbara

**Comment by galgeek** [comment]
Hi @whimboo!

I've updated.

Thank you again for all your help!

**Comment by whimboo** [comment]
This looks good to me. I rebased the commits and merged the branch as https://github.com/mozilla/testdaybot/commit/e38ec3ae5b32984fe4bf26e90637cdf6cb40f077.
